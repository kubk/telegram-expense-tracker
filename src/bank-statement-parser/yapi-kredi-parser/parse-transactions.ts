import * as pdfParse from 'pdf-parse';
import { assembleTransactions } from './assemble-transactions';
import { ITEM_SEPARATOR, ROW_SEPARATOR } from './constants';

const options = {
  normalizeWhitespace: true,
  disableCombineTextItems: true,
};

// https://www.npmjs.com/package/pdf-parse#extend
const renderPage: pdfParse.Options['pagerender'] = (pageData) => {
  return pageData
    .getTextContent(options)
    .then((textContent: { items: any }) => {
      let lastY = undefined;
      let text = '';
      for (const item of textContent.items) {
        if (lastY == item.transform[5] || !lastY) {
          text += ITEM_SEPARATOR + item.str;
        } else {
          text += ROW_SEPARATOR + item.str;
        }
        lastY = item.transform[5];
      }
      return text;
    });
};

export const parseTransactions = async (pdf: Buffer) => {
  const data = await pdfParse(pdf, { pagerender: renderPage });
  const result = data.text.split(ROW_SEPARATOR);

  return assembleTransactions(result);
};
