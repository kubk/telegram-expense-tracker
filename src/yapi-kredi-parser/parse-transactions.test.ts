import { existsSync, readFileSync } from 'fs';
import { parseTransactions } from './parse-transactions';

const table = [
  {
    pdfPath: '/Users/egor/Downloads/at1.pdf',
    testName: 'it parses one page PDF',
  },
  {
    pdfPath: '/Users/egor/Downloads/march.pdf',
    testName: 'it parses multi page page PDF',
  },
];

table.forEach((testCase) => {
  // Jest doesn't allow skipping tests programmatically, that's why we do it here with the loop
  const testOrSkip = existsSync(testCase.pdfPath) ? test : test.skip;

  testOrSkip(testCase.testName, async () => {
    const path = testCase.pdfPath;
    const dataBuffer = readFileSync(path);

    expect(await parseTransactions(dataBuffer)).toMatchSnapshot();
  });
});
