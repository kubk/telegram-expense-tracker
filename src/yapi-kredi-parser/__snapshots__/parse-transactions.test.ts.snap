// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it parses multi page page PDF 1`] = `
Array [
  Object {
    "amount": -10969,
    "createdAt": 2022-03-31T14:37:43.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 3103",
  },
  Object {
    "amount": -4363,
    "createdAt": 2022-03-31T07:35:49.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 3103",
  },
  Object {
    "amount": -5588,
    "createdAt": 2022-03-31T07:26:48.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ OZ BALLI MANAV 6254 3103",
  },
  Object {
    "amount": -4100,
    "createdAt": 2022-03-31T07:19:54.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ ALANYA TAVUKCUL 6254 | 3103",
  },
  Object {
    "amount": -3550,
    "createdAt": 2022-03-31T07:14:03.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ ERDEM ECZANESI 6254 3103",
  },
  Object {
    "amount": -53000,
    "createdAt": 2022-03-30T16:58:54.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ SYEDRA @ZEL SA¤ 6254 3003",
  },
  Object {
    "amount": -7210,
    "createdAt": 2022-03-30T15:36:24.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 3003",
  },
  Object {
    "amount": -5000,
    "createdAt": 2022-03-30T08:30:11.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ ALANYA TAVUKCUL 6254 | 3003",
  },
  Object {
    "amount": -2000,
    "createdAt": 2022-03-30T08:24:24.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ ERDEM ECZANESI 6254 3003",
  },
  Object {
    "amount": -3000,
    "createdAt": 2022-03-30T08:18:42.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ ANADOLU ECZANES 6254 | 3003",
  },
  Object {
    "amount": -6097,
    "createdAt": 2022-03-29T17:17:35.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 2903",
  },
  Object {
    "amount": -7303,
    "createdAt": 2022-03-29T17:07:46.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ OZ BALLI MANAV 6254 2903",
  },
  Object {
    "amount": -6490,
    "createdAt": 2022-03-28T18:34:36.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 2803",
  },
  Object {
    "amount": -12500,
    "createdAt": 2022-03-28T07:23:03.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ OZ BALLI MANAV 6254 2803",
  },
  Object {
    "amount": -6200,
    "createdAt": 2022-03-28T07:12:45.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 2803",
  },
  Object {
    "amount": -670,
    "createdAt": 2022-03-27T15:55:20.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ TOPLU TASIMA GE 6254 2703",
  },
  Object {
    "amount": -670,
    "createdAt": 2022-03-27T15:55:19.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ TOPLU TASIMA GE 6254 2703",
  },
  Object {
    "amount": -21300,
    "createdAt": 2022-03-27T15:46:29.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ SEYYAH DONER LO 6254 | 2703",
  },
  Object {
    "amount": -14347,
    "createdAt": 2022-03-27T14:58:30.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS WATSONS-276-ALANYUM 6254 | 2703",
  },
  Object {
    "amount": -904,
    "createdAt": 2022-03-27T14:32:20.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS OBA ALAN 6254 | 2703",
  },
  Object {
    "amount": -670,
    "createdAt": 2022-03-27T11:38:49.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ TOPLU TASIMA GE 6254 2703",
  },
  Object {
    "amount": -670,
    "createdAt": 2022-03-27T11:38:43.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ TOPLU TASIMA GE 6254 2703",
  },
  Object {
    "amount": -7621,
    "createdAt": 2022-03-26T16:35:17.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 2603",
  },
  Object {
    "amount": 700000,
    "createdAt": 2022-03-26T15:26:31.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "Para Yatırma ATM 104856 176752",
  },
  Object {
    "amount": -2000,
    "createdAt": 2022-03-26T09:59:36.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ AKSOY SIMIT 6254 2603",
  },
  Object {
    "amount": -8690,
    "createdAt": 2022-03-25T18:55:27.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ OZ BALLI MANAV 6254 2503",
  },
  Object {
    "amount": -24200,
    "createdAt": 2022-03-25T15:40:37.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ SOUL COFFEE 6254 2503",
  },
  Object {
    "amount": -538,
    "createdAt": 2022-03-25T11:07:40.000Z,
    "currency": "TRY",
    "info": "Other / Internet - Mobile",
    "title": "DOVIZ ISLEM KMV",
  },
  Object {
    "amount": -269007,
    "createdAt": 2022-03-25T11:07:40.000Z,
    "currency": "TRY",
    "info": "Foreign Exchange / Internet - Mobile",
    "title": "USD MOBIL DOVIZ ALIS 14.94484",
  },
  Object {
    "amount": -6300,
    "createdAt": 2022-03-24T11:33:45.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ 7 WAYS 6254 2403",
  },
  Object {
    "amount": -5900,
    "createdAt": 2022-03-24T11:26:21.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ AKSOY SIMIT 6254 2403",
  },
  Object {
    "amount": -12549,
    "createdAt": 2022-03-24T09:16:02.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MAVI MARKET 6254 2403",
  },
  Object {
    "amount": -3576,
    "createdAt": 2022-03-23T08:18:22.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 2303",
  },
  Object {
    "amount": -76500,
    "createdAt": 2022-03-22T14:01:26.000Z,
    "currency": "TRY",
    "info": "Money Transfer / Internet - Mobile",
    "title": "GELEN HAVALE -EL ZARA MUSLIMOVA ",
  },
  Object {
    "amount": -44335,
    "createdAt": 2022-03-22T09:29:57.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "EPOS FLAT IO MUSIC NOTAT 6254 2203",
  },
  Object {
    "amount": -7593,
    "createdAt": 2022-03-21T17:16:11.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ OZ BALLI MANAV 6254 2103",
  },
  Object {
    "amount": -4700,
    "createdAt": 2022-03-21T16:49:21.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ ATOM MASTER 6254 2103",
  },
  Object {
    "amount": 104800,
    "createdAt": 2022-03-21T15:13:53.000Z,
    "currency": "TRY",
    "info": "Money Transfer / Other",
    "title": "GELEN FAST - İSMAİL KESKİN - İSMAİL | KESKİN tarafından aktarılan",
  },
  Object {
    "amount": -14800,
    "createdAt": 2022-03-20T13:18:14.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ ALICAN PIDE 6254 2003",
  },
  Object {
    "amount": -13592,
    "createdAt": 2022-03-19T09:16:13.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 1903",
  },
  Object {
    "amount": -6418,
    "createdAt": 2022-03-18T08:12:39.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "EPOS PAYPAL *GERMANCOMPU 6254 | 1803",
  },
  Object {
    "amount": 106,
    "createdAt": 2022-03-18T08:12:21.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "EPOS PAYPAL 6254 HATA",
  },
  Object {
    "amount": -106,
    "createdAt": 2022-03-18T08:12:21.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "EPOS PAYPAL 6254 1803HATA",
  },
  Object {
    "amount": -7000,
    "createdAt": 2022-03-18T07:55:02.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ OZ BALLI MANAV 6254 1803",
  },
  Object {
    "amount": -13512,
    "createdAt": 2022-03-18T07:40:27.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ ERDEM ECZANESI 6254 1803",
  },
  Object {
    "amount": -19105,
    "createdAt": 2022-03-18T07:29:17.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 1803",
  },
  Object {
    "amount": -8,
    "createdAt": 2022-03-17T09:09:11.000Z,
    "currency": "TRY",
    "info": "Other / Internet - Mobile",
    "title": "BSMV ELEKTRONİK FON TRANSFERİ | (EFT) ÜCRETİ-FAST/Anna Zolotova",
  },
  Object {
    "amount": -155,
    "createdAt": 2022-03-17T09:09:11.000Z,
    "currency": "TRY",
    "info": "Other / Internet - Mobile",
    "title": "ELEKTRONİK FON TRANSFERİ (EFT) | ÜCRETİ-FAST/Anna Zolotova",
  },
  Object {
    "amount": -8100,
    "createdAt": 2022-03-17T09:09:11.000Z,
    "currency": "TRY",
    "info": "Money Transfer / Internet - Mobile",
    "title": "GİDEN FAST - Anna Zolotova",
  },
  Object {
    "amount": -1476,
    "createdAt": 2022-03-17T06:48:18.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 1703",
  },
  Object {
    "amount": -6000,
    "createdAt": 2022-03-17T06:40:48.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ OZ BALLI MANAV 6254 1703",
  },
  Object {
    "amount": -12,
    "createdAt": 2022-03-16T18:16:54.000Z,
    "currency": "TRY",
    "info": "Other / Internet - Mobile",
    "title": "BSMV ELEKTRONİK FON TRANSFERİ | (EFT) ÜCRETİ-FAST/Anna Zolotova",
  },
  Object {
    "amount": -232,
    "createdAt": 2022-03-16T18:16:54.000Z,
    "currency": "TRY",
    "info": "Other / Internet - Mobile",
    "title": "ELEKTRONİK FON TRANSFERİ (EFT) | ÜCRETİ-FAST/Anna Zolotova",
  },
  Object {
    "amount": -29600,
    "createdAt": 2022-03-16T18:16:54.000Z,
    "currency": "TRY",
    "info": "Money Transfer / Internet - Mobile",
    "title": "GİDEN FAST - Anna Zolotova",
  },
  Object {
    "amount": -35000,
    "createdAt": 2022-03-15T16:38:23.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ IK RA AYAKKABI 6254 1503",
  },
  Object {
    "amount": -9170,
    "createdAt": 2022-03-14T17:30:49.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "EPOS TRENDYOL.COM 6254 1403",
  },
  Object {
    "amount": -7388,
    "createdAt": 2022-03-14T17:21:55.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 1403",
  },
  Object {
    "amount": -15879,
    "createdAt": 2022-03-14T13:59:54.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 1403",
  },
  Object {
    "amount": -12480,
    "createdAt": 2022-03-14T13:45:02.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ SAFRAN BAHARAT 6254 1403",
  },
  Object {
    "amount": 500000,
    "createdAt": 2022-03-14T13:35:51.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "Para Yatırma ATM 104949 205599",
  },
  Object {
    "amount": -2100,
    "createdAt": 2022-03-14T13:25:48.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ AKSOY SIMIT 6254 1403",
  },
  Object {
    "amount": -670,
    "createdAt": 2022-03-14T09:07:41.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ TOPLU TASIMA GE 6254 1403",
  },
  Object {
    "amount": -1500,
    "createdAt": 2022-03-14T08:35:09.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ EKREM COSKUN 6254 1403",
  },
  Object {
    "amount": 5000,
    "createdAt": 2022-03-14T08:32:40.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "Para Yatırma ATM 100767 263939",
  },
]
`;

exports[`it parses one page PDF 1`] = `
Array [
  Object {
    "amount": -2788,
    "createdAt": 2022-04-04T13:58:35.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 0404",
  },
  Object {
    "amount": -5598,
    "createdAt": 2022-04-04T13:50:20.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MJET MAHMUTLAR 6254 | 0404",
  },
  Object {
    "amount": -2100,
    "createdAt": 2022-04-04T13:29:25.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ ATOM MASTER 6254 0404",
  },
  Object {
    "amount": -18732,
    "createdAt": 2022-04-03T20:25:59.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "EPOS IYZICO AMAZO*AMAZON 6254 | 0304",
  },
  Object {
    "amount": -4502,
    "createdAt": 2022-04-03T13:21:13.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 0304",
  },
  Object {
    "amount": -10517,
    "createdAt": 2022-04-02T09:00:34.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MIGROS ATATURK 6254 0204",
  },
  Object {
    "amount": -16000,
    "createdAt": 2022-04-02T08:55:16.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ CRYPTO ILETISIM 6254 0204",
  },
  Object {
    "amount": -9154,
    "createdAt": 2022-04-02T08:45:32.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ OZ BALLI MANAV 6254 0204",
  },
  Object {
    "amount": -3900,
    "createdAt": 2022-04-02T08:38:12.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ BIM-50. YIL -J7 6254 0204",
  },
  Object {
    "amount": -3900,
    "createdAt": 2022-04-02T08:37:38.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ BIM-50. YIL -J7 6254 0204",
  },
  Object {
    "amount": -2189,
    "createdAt": 2022-04-02T08:32:28.000Z,
    "currency": "TRY",
    "info": "Other / Other",
    "title": "POS TMSZ MAVI MARKET 6254 0204",
  },
  Object {
    "amount": -651,
    "createdAt": 2022-04-01T10:47:03.000Z,
    "currency": "TRY",
    "info": "Other / Internet - Mobile",
    "title": "DOVIZ ISLEM KMV",
  },
  Object {
    "amount": -325524,
    "createdAt": 2022-04-01T10:47:03.000Z,
    "currency": "TRY",
    "info": "Foreign Exchange / Internet - Mobile",
    "title": "USD MOBIL DOVIZ ALIS 14.79654",
  },
  Object {
    "amount": 415034,
    "createdAt": 2022-04-01T07:24:02.000Z,
    "currency": "TRY",
    "info": "Money Transfer / Other",
    "title": "GELEN EFT - BİRLEŞİK ÖDEME | HİZMETLERİ VE ELEKTRONİK PARA A.Ş. - | PYS537336797846 - transfer.",
  },
]
`;
